src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.IO.Pipelines
+{
+    /// <summary>
+    /// Default <see cref="PipeWriter"/> and <see cref="PipeReader"/> implementation.
+    /// </summary>
+    public sealed class Pipe
+    {
+        private const int SegmentPoolSize = 16;
```

Comment:
Should this be configurable in PipeOptions?

Reply:
We decided to try const first with the idea that if you have more you can increase minimum block size and have larger blocks.

Reply:
I think we should make it into an option, but it's ok to punt for later.